snuba/web/db_query.py

```diff
@@ -55,6 +57,152 @@ def update_query_metadata_and_stats(
     return stats
 
 
+def execute_query(
+    request: Request,
+    query: ClickhouseQuery,
+    timer: Timer,
+    stats: MutableMapping[str, Any],
+    query_settings: MutableMapping[str, Any],
+    query_id: Optional[str],
+) -> Result:
+    """
+    Execute a query and return a result.
+    """
+    # Experiment, if we are going to grab more than X columns worth of data,
+    # don't use uncompressed_cache in ClickHouse.
+    uc_max = state.get_config("uncompressed_cache_max_cols", 5)
+    if len(request.query.get_all_referenced_columns()) > uc_max:
+        query_settings["use_uncompressed_cache"] = 0
+
+    # Force query to use the first shard replica, which
+    # should have synchronously received any cluster writes
+    # before this query is run.
+    consistent = request.settings.get_consistent()
+    stats["consistent"] = consistent
+    if consistent:
+        query_settings["load_balancing"] = "in_order"
+        query_settings["max_threads"] = 1
+
+    result = reader.execute(
+        query,
+        query_settings,
+        query_id=query_id,
+        with_totals=request.query.has_totals(),
+    )
+
+    timer.mark("execute")
+    stats.update(
+        {"result_rows": len(result["data"]), "result_cols": len(result["meta"])}
+    )
+
+    return result
+
+
+def execute_query_with_rate_limits(
+    request: Request,
+    query: ClickhouseQuery,
+    timer: Timer,
+    stats: MutableMapping[str, Any],
+    query_settings: MutableMapping[str, Any],
+    query_id: Optional[str],
+) -> Result:
+    # XXX: We should consider moving this that it applies to the logical query,
+    # not the physical query.
+    with RateLimitAggregator(
+        request.settings.get_rate_limit_params()
+    ) as rate_limit_stats_container:
+        stats.update(rate_limit_stats_container.to_dict())
+        timer.mark("rate_limit")
+
+        project_rate_limit_stats = rate_limit_stats_container.get_stats(
+            PROJECT_RATE_LIMIT_NAME
+        )
+
+        if (
+            "max_threads" in query_settings
+            and project_rate_limit_stats is not None
+            and project_rate_limit_stats.concurrent > 1
+        ):
+            maxt = query_settings["max_threads"]
+            query_settings["max_threads"] = max(
+                1, maxt - project_rate_limit_stats.concurrent + 1
+            )
+
+        return execute_query(request, query, timer, stats, query_settings, query_id)
+
+
+def execute_query_with_caching(
```

Comment:
This will likely be replaced by #882, but will need to be able to exist alongside it for a while if we want to have a runtime switch for the new implementation.